### ISSUE TITLE
Java code generation does not compile when combining @includes and inline fragments

### ISSUE THREAD

**Issue by morganlutz** [apollographql/apollo-android#1884]
**Summary**
I am trying to upgrade from version `1.0.0` to version `1.2.2`. I am not sure if this is a bug or if there was a change made and I just need to define my fragment differently. Potentially the fragment has always been defined incorrectly and the new version exposes that. The java generated code for the given example no longer compiles.

**Description**
Schema: https://api.kivaws.org/graphql
Fragment definition:
```
fragment LoanFragment on LoanBasic {
    id
    status
    name
    use
    loanAmount
    plannedExpirationDate
    loanFundraisingInfo {
        fundedAmount
        reservedAmount
    }
    image {
        url(presetSize: loan_retina)
    }
    geocode {
        country {
            isoCode
            name
        }
    }
    ... on LoanBasic @include(if: $withDetails) {
        raisedDate
        description
        whySpecial
        activity {
            name
        }
        sector {
            id
            name
        }
        hasCurrencyExchangeLossLenders
        terms {
            lenderRepaymentTerm
        }
        repaymentInterval
        paidAmount
        userProperties {
            lentTo
        }
        isMatchable
        matcherName
    }
    ... on LoanPartner @include(if: $withDetails) {
        partnerName
        partner {
            id
            image {
                url(customSize: "h300")
            }
            loanAlertText
            chargesFeesInterest
            riskRating
        }
    }
}
```

Java code generation for `LoanPartner`:

In version `1.0.0`:

```
  class AsLoanPartner implements LoanFragment {
    static final ResponseField[] $responseFields = {
      ResponseField.forString("__typename", "__typename", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forInt("id", "id", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forString("status", "status", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forString("name", "name", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forString("use", "use", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forCustomType("loanAmount", "loanAmount", null, false, CustomType.MONEY, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forCustomType("plannedExpirationDate", "plannedExpirationDate", null, true, CustomType.DATE, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forObject("loanFundraisingInfo", "loanFundraisingInfo", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forObject("image", "image", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forObject("geocode", "geocode", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forCustomType("raisedDate", "raisedDate", null, true, CustomType.DATE, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false))),
      ResponseField.forString("description", "description", null, true, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false))),
      ResponseField.forString("whySpecial", "whySpecial", null, true, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false))),
      ResponseField.forObject("activity", "activity", null, true, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false))),
      ResponseField.forObject("sector", "sector", null, true, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false))),
      ResponseField.forBoolean("hasCurrencyExchangeLossLenders", "hasCurrencyExchangeLossLenders", null, true, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false))),
      ResponseField.forObject("terms", "terms", null, true, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false))),
      ResponseField.forString("repaymentInterval", "repaymentInterval", null, true, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false))),
      ResponseField.forCustomType("paidAmount", "paidAmount", null, true, CustomType.MONEY, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false))),
      ResponseField.forObject("userProperties", "userProperties", null, true, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false))),
      ResponseField.forBoolean("isMatchable", "isMatchable", null, true, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false))),
      ResponseField.forString("matcherName", "matcherName", null, true, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false))),
      ResponseField.forString("partnerName", "partnerName", null, true, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false))),
      ResponseField.forObject("partner", "partner", null, true, Arrays.<ResponseField.Condition>asList(ResponseField.Condition.booleanCondition("withDetails", false)))
    };
```

In version `1.2.2.`:
```
  class AsLoanPartner implements LoanFragment {
    static final ResponseField[] $responseFields = {
      ResponseField.forString("__typename", "__typename", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forInt("id", "id", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forString("status", "status", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forString("name", "name", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forString("use", "use", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forCustomType("loanAmount", "loanAmount", null, false, CustomType.MONEY, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forCustomType("plannedExpirationDate", "plannedExpirationDate", null, true, CustomType.DATE, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forObject("loanFundraisingInfo", "loanFundraisingInfo", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forObject("image", "image", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forObject("geocode", "geocode", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forString("partnerName", "partnerName", null, false, Collections.<ResponseField.Condition>emptyList()),
      ResponseField.forObject("partner", "partner", null, true, Collections.<ResponseField.Condition>emptyList())
    };
```

The compilation errors are due to missing field generation:
`Class `AsLoanPartner` must either be declared abstract or implement abstract method...`

The goal is to include some fields that are only present if the fragment implements `LoanPartner`,  and return `null` for those fields if it does not. I would like to keep the same `.graphql` files for both Android and iOS, and this works in iOS.

**Version**
`1.2.2`

**Comment by sav007** [comment]
Sorry, could you please elaborate more on what fields are missing?

**Comment by morganlutz** [comment]
@sav007 Every field included in 
`... on LoanBasic @include(if: $withDetails) {` 

which is *not* included in 
`... on LoanPartner @include(if: $withDetails) {`

I get errors that these methods are not implemented in `LoanFragment.AsLoanPartner`:

```
raisedDate(): Date
description(): String
whySpecial(): String
activity(): Activity
sector(): Sector
hasCurrencyExchangeLossLenders(): boolean
terms(): Terms
repaymentInterval(): String
paidAmount(): String
userProperties(): UserProperties
isMatchable(): boolean
matcherName(): String
```
I removed `@includes` to fix #1883 which also fixed this issue, so we were able to upgrade. ⭐️ 

This works fine:

```
fragment LoanDetailsFragment on LoanBasic {
    raisedDate                             
    description                            
    whySpecial                             
    sector {                               
        id                                 
        name                               
    }                                      
    hasCurrencyExchangeLossLenders         
    terms {                                
        lenderRepaymentTerm                
    }                                      
    repaymentInterval                      
    paidAmount                             
    userProperties {                       
        lentTo                             
    }                                      
    ... on LoanPartner {                   
        partner {                          
            id                             
            name                           
            image {                        
                url(customSize: "h300")    
            }                              
            loanAlertText                  
            chargesFeesInterest            
            riskRating                     
        }                                  
    }                                      
}
```
